components/salsa-macro-rules/src/setup_input_struct.rs

```diff
@@ -204,6 +229,72 @@ macro_rules! setup_input_struct {
                     })
                 }
             }
+
+            impl $zalsa_struct::HasBuilder for $Struct {
+                type Builder = builder::$Builder;
+            }
+
+            mod builder {
+                use super::*;
+
+                use salsa::plumbing as $zalsa;
+                use $zalsa::input as $zalsa_struct;
+
+                // These are standalone functions instead of methods on `Builder` to prevent
+                // that the enclosing module can call them.
+                pub(super) fn new_builder($($field_id: $field_ty),*) -> $Builder {
+                    $Builder {
+                        fields: ($($field_id,)*),
+                        durabilities: [$zalsa::Durability::default(); $N],
+                    }
+                }
+
+                pub(super) fn builder_into_inner(builder: $Builder, revision: $zalsa::Revision) -> (($($field_ty,)*), $zalsa::Array<$zalsa::Stamp, $N>) {
+                    let stamps = $zalsa::Array::new([
+                        $($zalsa::stamp(revision, builder.durabilities[$field_index])),*
+                    ]);
+
+                    (builder.fields, stamps)
+                }
+
+                pub struct $Builder {
+                    /// The field values.
+                    fields: ($($field_ty,)*),
+
+                    /// The durabilities per field.
+                    durabilities: [$zalsa::Durability; $N],
+                }
+
+                impl $Builder {
+                    /// Sets the durability of all fields.
+                    ///
+                    /// Overrides any previously set durabilities.
+                    pub fn durability(mut self, durability: $zalsa::Durability) -> Self {
+                        self.durabilities = [durability; $N];
+                        self
+                    }
+
+                    $(
+                        /// Sets the value of the field `$field_id`.
+                        #[must_use]
+                        pub fn $field_id(mut self, value: $field_ty) -> Self
+                        {
+                            self.fields.$field_index = value;
+                            self
+                        }
+                    )*
```

Comment:
This also generates setters for required fields which is redundant because they were already provided when calling `builder`. However, not adding them is slightly more complicated and probably requires explicitly passing an `optional_fields` structure to the macro with the `id`, `ty` and `index` of each optional field.